nemo_skills/code_execution/local_sandbox/local_sandbox_server.py

```diff
@@ -129,31 +141,77 @@ def execute_ipython_session(generated_code, session_id, traceback_verbosity="Pla
         shell = session_data["shell"]
         shell.InteractiveTB.set_mode(mode=traceback_verbosity)
 
-        # Capture stdout/stderr
-        try:
-            with io.capture_output() as captured:
-                result = shell.run_cell(generated_code)
-            stdout_result = captured.stdout
-            stderr_result = captured.stderr
-            has_error = result.error_before_exec or result.error_in_exec
-            process_status = "completed" if not has_error else "error"
-
-        except Exception as e:
-            process_status = "error"
-            stdout_result = captured.stdout if "captured" in locals() else ""
-            stderr_extra = f"\n{type(e).__name__}: {e}"
-            stderr_result = (captured.stderr if "captured" in locals() else "") + stderr_extra
-            stdout_result += stderr_result
-            stderr_result = ""
+        # Use threading to implement timeout for IPython execution
+        execution_result = {}
+        execution_error = {}
 
-        return {
-            "process_status": process_status,
-            "stdout": postprocess_output(stdout_result, traceback_verbosity),
-            "stderr": postprocess_output(stderr_result, traceback_verbosity),
-            "new_session_created": new_session_created,
-        }
+        def execute_with_capture():
+            """Execute the code in a separate thread to enable timeout"""
+            try:
+                with io.capture_output() as captured:
+                    result = shell.run_cell(generated_code)
+                execution_result["stdout"] = captured.stdout
+                execution_result["stderr"] = captured.stderr
+                execution_result["has_error"] = result.error_before_exec or result.error_in_exec
+                execution_result["success"] = True
+            except Exception as e:
+                execution_error["exception"] = e
+                execution_error["stdout"] = captured.stdout if "captured" in locals() else ""
+                execution_error["stderr"] = (
+                    captured.stderr if "captured" in locals() else ""
+                ) + f"\n{type(e).__name__}: {e}"
+
+        # Start execution in a separate thread
+        execution_thread = threading.Thread(target=execute_with_capture)
+        execution_thread.daemon = True  # Allow main thread to exit even if this is still running
+        execution_thread.start()
+
+        # Wait for completion or timeout
+        execution_thread.join(timeout=timeout)
```

Comment:
what happens to this thread after we return from this function? Does it keep running or nginx cleans it up somehow?

Reply:
yeah this is a good call out, ti will keep running until it completes. I am working on a fix to better manage canceling the process running the cell.